### PR TITLE
[C++] Avoid using dynamic_cast where possible by using hand rolled RTTI

### PR DESCRIPTION
**Pull request by jcking**
`dynamic_cast` in hot code paths is much slower than using hand rolled RTTI using an enum. CPU profiling shows that `dynamic_cast` is one of the single most expensive calls. This change switches to using a hand rolled RTTI-like implementation using an enum.

### CODE REVIEW

**Comment by QuantamHD** [comment]
@parrt I'm going to upload my draft PR for posterity, and include the performance numbers.

**Comment by QuantamHD** [comment]
I'm waiting on Google to approve the DCO in this repo, but here's my version https://github.com/antlr/antlr4/compare/dev...QuantamHD:surelog_perf?expand=1

It results in about a 25% performance uplift in walltime, and 11% reduction in profile cycles.

Before Changes
image


After Changes
image

After the changes strcmp and dynamic_cast are nowhere near the top of the profile. I think @jcking's changes are great. I made mine more targeted by implementing specific is* functions for hot paths whereas Justin's is far more general. Either way removing RTTI in hot paths is a must.

File: runtime/Cpp/runtime/src/atn/SingletonPredictionContext.cpp [hunk 1]
```diff
@@ -45,11 +49,11 @@ bool SingletonPredictionContext::operator == (const PredictionContext &o) const
     return true;
   }
 
-  const SingletonPredictionContext *other = dynamic_cast<const SingletonPredictionContext*>(&o);
-  if (other == nullptr) {
+  if (o.getContextType() != PredictionContextType::EMPTY && o.getContextType() != PredictionContextType::SINGLETON) {
```
Comment (mike-lischke) [inline runtime/Cpp/runtime/src/atn/SingletonPredictionContext.cpp:52]: Why do we need the extra test for ::EMPTY here?

**Review by mike-lischke — COMMENTED** [review]
Looks good, just a small question for you @jcking .

File: runtime/Cpp/runtime/src/atn/PredictionContext.cpp [hunk 2]
```diff
@@ -389,11 +389,11 @@ std::string PredictionContext::toDOTString(const Ref<const PredictionContext> &c
   });
 
   for (const auto &current : nodes) {
-    if (is<const SingletonPredictionContext>(current)) {
+    if (current->getContextType() == PredictionContextType::SINGLETON || current->getContextType() == PredictionContextType::EMPTY) {
```
Comment (hzeller) [inline runtime/Cpp/runtime/src/atn/PredictionContext.cpp:392]: Since we're calling the virtual `getContextType()` multiple times in this block (here and below, in total 3 times), it might be worthwhile calling it only once, remembering the result and do the cheap comparisons without the virtual method call overhead multiple times. Picking this as an example, there are also the other places that might need to call virtual functions multiple times.

```c++
  for (const auto &current : nodes) {
     const PredictionContextType contextType = current->getContextType();
      if (contextType == PredictionContextType::SINGLETON || contextType == PredictionContextType::EMPTY) {
            //...
            if (contextType == PredictionContextType::EMPTY) {
              // ...
            }
       }
  }
```

File: runtime/Cpp/runtime/src/atn/ATNStateType.h [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    TOKEN_START = 6,
+    RULE_STOP = 7,
+    BLOCK_END = 8,
+    STAR_LOOP_BACK = 9,
+    STAR_LOOP_ENTRY = 10,
+    PLUS_LOOP_BACK = 11,
+    LOOP_END = 12,
+  };
+
+  std::string atnStateTypeName(ATNStateType atnStateType);
```
Comment (hzeller) [inline runtime/Cpp/runtime/src/atn/ATNStateType.h:33]: Not necessarily now (not sure what the current baseline requirements are, is antlr already compiling with c++17?), but longer term, using `std::string_view` here is probably a good idea.

File: runtime/Cpp/runtime/src/atn/SingletonPredictionContext.h [hunk 4]
```diff
@@ -25,6 +25,8 @@ namespace atn {
 
     static Ref<const SingletonPredictionContext> create(Ref<const PredictionContext> parent, size_t returnState);
 
+    PredictionContextType getContextType() const override;
```
Comment (hzeller) [inline runtime/Cpp/runtime/src/atn/SingletonPredictionContext.h:28]: For all classes that are not extended anymore, it would probably be good to use `final` instead of `override`.
Not only is this a good documentation for the reader of the code, but the compiler can also use that for devirtualization if it otherwise can know which class it is dealing with.

Comment (jcking) [reply runtime/Cpp/runtime/src/atn/SingletonPredictionContext.h:28]: Unfortunately SingletonPredictionContext is inherited by EmptyPredictionContext, so it cannot be final. However, whether or not we need PredictionContextType::EMPTY I will have to look.

File: runtime/Cpp/runtime/src/atn/ATNStateType.h (hunk 3 above)
Comment (jcking) [reply runtime/Cpp/runtime/src/atn/ATNStateType.h:33]: We could use string_view, but then we cannot ever dynamically allocate. For example, if we wanted to do "INVALID(%d)".

File: runtime/Cpp/runtime/src/atn/PredictionContext.cpp (hunk 2 above)
Comment (jcking) [reply runtime/Cpp/runtime/src/atn/PredictionContext.cpp]: Fixed.

File: runtime/Cpp/runtime/src/atn/SingletonPredictionContext.cpp (hunk 1 above)
Comment (jcking) [reply runtime/Cpp/runtime/src/atn/SingletonPredictionContext.cpp]: Fixed, we do not now. I removed EmptyPredictionContext as its not useful and can be implemented purely in terms of SIngletonPredictionContext. It also avoids people instantiating EmptyPredictionContext, since some code uses pointer equality to the PredictionContext::EMPTY.

File: runtime/Cpp/runtime/src/atn/SingletonPredictionContext.h (hunk 4 above)
Comment (jcking) [reply runtime/Cpp/runtime/src/atn/SingletonPredictionContext.h:28]: I was able to remove EmptyPredictionContext and make SingletonPredictionContext final. However it won't matter much, as most invocations use the base class rather than the derived. So the compiler cannot do much.

Comment (hzeller) [reply runtime/Cpp/runtime/src/atn/SingletonPredictionContext.h:28]: I usually find it most useful as documentation for the reader of the code, actually.

File: runtime/Cpp/runtime/src/atn/ATNStateType.h (hunk 3 above)
Comment (hzeller) [reply runtime/Cpp/runtime/src/atn/ATNStateType.h:33]: Given that `ATNStateType` is an enum class it would be very hard to get an invalid value in there. So I'd prefer performance over new string allocations to be prepared for an unlikely force-typecast any time.

File: runtime/Cpp/runtime/src/atn/ATNStateType.cpp [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+      return "BLOCK_END";
+    case ATNStateType::STAR_LOOP_BACK:
+      return "STAR_LOOP_BACK";
+    case ATNStateType::STAR_LOOP_ENTRY:
+      return "STAR_LOOP_ENTRY";
+    case ATNStateType::PLUS_LOOP_BACK:
+      return "PLUS_LOOP_BACK";
+    case ATNStateType::LOOP_END:
+      return "LOOP_END";
+    default:
```
Comment (hzeller) [inline runtime/Cpp/runtime/src/atn/ATNStateType.cpp:29]: I suggest not having a default case. That way, the compiler automatically will warn when there is a new enum value that is not handled. In general whenever there are default cases in switch/case statements that are supposed to handle all cases of an enum, we should consider removing them to have the compiler help us check for correctness.

So I suggest to move the `return "Invalid";` to the end, outside the switch/case.

File: runtime/Cpp/runtime/src/atn/TransitionType.cpp [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+      return "ACTION";
+    case TransitionType::SET:
+      return "SET";
+    case TransitionType::NOT_SET:
+      return "NOT_SET";
+    case TransitionType::WILDCARD:
+      return "WILDCARD";
+    case TransitionType::PRECEDENCE:
+      return "PRECEDENCE";
+    default:
```
Comment (hzeller) [inline runtime/Cpp/runtime/src/atn/TransitionType.cpp:25]: Similar suggestion here with the default case: Don't have a default case here, but move the `return "INVALID";` to the end of the function.

File: runtime/Cpp/runtime/src/atn/ATNStateType.h (hunk 3 above)
Comment (jcking) [reply runtime/Cpp/runtime/src/atn/ATNStateType.h:33]: atnStateTypeName and transitionTypeName are only used by toString methods, which are not called anywhere in the ANTLR runtime except for debugging, which is off by default. So performance is not an issue in this case, unless the user is calling toString in which cause I would suggest not unless debugging because a bunch of them are not performant.

**Review by mike-lischke — APPROVED** [review]
From my standpoint the patch is fine, but @hzeller has a few good points here, which might be worth to be considered.

File: runtime/Cpp/runtime/src/atn/TransitionType.cpp (hunk 6 above)
Comment (jcking) [reply runtime/Cpp/runtime/src/atn/TransitionType.cpp]: Ugh, didn't public comment. I moved it to the end, but used UNKNOWN. ATNStateType uses INVALID.

File: runtime/Cpp/runtime/src/atn/ATNStateType.cpp (hunk 5 above)
Comment (jcking) [reply runtime/Cpp/runtime/src/atn/ATNStateType.cpp]: Same as other, fixed.

**Comment by jcking** [comment]
> From my standpoint the patch is fine, but @hzeller has a few good points here, which might be worth to be considered.

Addressed the switch cases. @parrt we should be good assuming tests for C++ are successful.

**Comment by parrt** [comment]
Ok, great! I'll address the php failing issue later.
